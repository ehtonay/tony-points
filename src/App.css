
html {
	scroll-behavior: smooth;
}

body {
  font-family: Arial;
  font-size: 12px;
  background: lightskyblue;
  text-align: center;
}

input, button {
  padding: 10px;
  margin: 16px;
  border-radius: 5px;
  border: 1px solid #444;
}

button {
  width: 100px;
  cursor: pointer;
}

#errorMessage {
  color: red;
  font-weight: bold;
}

#loading {
  display: inline-block;
  width: 40px;
  height: 40px;
  border: 3px solid #00add8;
  border-radius: 50%;
  border-top-color: #fff;
  animation: spin 1s ease-in-out infinite;
  -webkit-animation: spin 1s ease-in-out infinite;
}
  
@keyframes spin {
  to { -webkit-transform: rotate(360deg); }
}

@-webkit-keyframes spin {
  to { -webkit-transform: rotate(360deg); }
}

#results {
  background: white;
  width: 50%;
  margin: 20px auto;
  padding: 20px;
  border-radius: 5px;
  box-shadow: 1px 2px 4px -1px #000;
}

.tile {
  background: white;
  margin: 10px;
  padding: 10px;
  box-shadow: 1px 2px 4px -1px #000;
}