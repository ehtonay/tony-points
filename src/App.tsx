import * as React from "react";
import { useState } from "react";
import { TaxBracket, TaxResponse } from "./types";
import { BASE_URL } from "./constants";
import { TaxBracketTile } from "./components/TaxBracketTile";
import { ErrorAlert } from "./components/ErrorAlert";
import './App.css';

/**
 * Marginal Taxation Rate Calculator App
 * @author Tony Tomarchio
 */
export const App = () => {
  const [taxYear, setTaxYear] = useState('');
  const [salary, setSalary] = useState('');
  const [taxBrackets, setTaxBrackets] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [hasErrors, setHasErrors] = useState('');

  // Would extract to custom hook for multiple currencies
  const currencyFormatter = new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' });

  const handleSubmit = async (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    e.preventDefault();
    setIsLoading(true);

    // Validate input - would usually use moment.js for date checks
    const income = parseInt(salary);
    if (taxYear.length !== 4 || isNaN(income)) {
      setHasErrors('Please use a valid year/income combination.')
      setIsLoading(false);
      return;
    }

    // Fetch tax brackets data for the year
    await fetch(`${BASE_URL}/${taxYear}`)
      .then(res => res.json())
      .then((res: TaxResponse) => {
        if (res.errors) {
          console.error(res.errors);
          setHasErrors(`Failed to fetch the tax bracket data for ${taxYear}`)
        } else {
        // Calculate taxes paid in each bracket and store value in the object
        res.tax_brackets?.forEach((taxBracket) => {
          if (income > taxBracket.min) {
            if (!taxBracket.max) {
              taxBracket.taxesPaid = (income - taxBracket.min) * taxBracket.rate;
            } else {
              if (income > taxBracket.max) {
                taxBracket.taxesPaid = (taxBracket.max - taxBracket.min) * taxBracket.rate;
              } else {
                taxBracket.taxesPaid = (income - taxBracket.min) * taxBracket.rate;
              }
            }
          }
        })
        // Sort brackets to display in ascending order for UI
        const sortedBrackets: TaxBracket[] = res.tax_brackets.sort((a, b) => {
          return a.min < b.min ? -1 : 1;
        });
        setTaxBrackets(sortedBrackets);
      }
    });
    setIsLoading(false);
  }

  const resetInput = () => {
    setSalary('');
    setTaxYear('');
    setTaxBrackets([]);
    setHasErrors('');
  }

  // Build out UI using sorted, massaged tax bracket data
  let totalTaxesPaid = 0;
  const taxBracketsView = taxBrackets.map(bracket => {
    // Made the choice to not display brackets with no taxes paid
    if (bracket.taxesPaid) {
      totalTaxesPaid += bracket.taxesPaid;
      const formattedTaxesPaid = currencyFormatter.format(bracket.taxesPaid);
      return (
        <TaxBracketTile
          min={currencyFormatter.format(bracket.min)}
          max={bracket.max && currencyFormatter.format(bracket.max)}
          taxesPaid={formattedTaxesPaid} key={bracket.min}
        />
      )
    }
    return null;
  });

  if (isLoading) return <div id="loading" />;

  // Render - usually would localize strings for translation (if needed)
  return taxBrackets.length > 0 ? (
    <>
      <h1>Results</h1>
      <div id='results'>
        <b>Your salary:</b> {currencyFormatter.format(parseInt(salary))}<br />
        <b>Tax Year:</b> {taxYear}<br />
        <b>Total taxes paid:</b> {currencyFormatter.format(totalTaxesPaid)}
      </div>
      <div style={{ display: 'flex', justifyContent: 'space-evenly' }}>
        {taxBracketsView}
      </div>
      <button onClick={resetInput}>Reset</button>
    </>
  ) : (
    <>
      <h1>Marginal Tax Bracket Calculator</h1>
      {hasErrors && <ErrorAlert errorMessage={hasErrors} />}
      <form>
        <input autoFocus type="text" onChange={e => setTaxYear(e.target.value)} placeholder="Enter a tax year" />
        <input type="text" onChange={e => setSalary(e.target.value)} placeholder="Enter your annual salary" />
        <div>
          <button type="submit" onClick={handleSubmit}>Submit</button>
        </div>
      </form>
    </>
  );
}
