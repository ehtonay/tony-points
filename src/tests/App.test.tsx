import * as React from "react";
import { act } from 'react-dom/test-utils';
import * as ReactDOM from "react-dom";
import { App } from "../App";

describe('App', function () {
    it('should display pass in number', function () {
        const container = document.createElement('div');
        document.body.appendChild(container);
        act(() => {
            ReactDOM.render(<App />, container);
        })

        // Basic state element assertions and snapshot
        const header = container.querySelector('h1');
        expect(header.textContent).toBe("Marginal Tax Bracket Calculator")
        expect(container).toMatchSnapshot();

        // Simulate change on input elements to add tax year and salary

        // Simulate click on submit button

        // Mock fetch to return valid data and assert elements and snapshot

        // Mock fetch to return invalid data and assert elements and snapshot
    });
});