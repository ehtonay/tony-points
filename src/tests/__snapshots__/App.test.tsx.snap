// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should display pass in number 1`] = `
<div>
  <h1>
    Marginal Tax Bracket Calculator
  </h1>
  
  <form>
    <input
      placeholder="Enter a tax year"
      type="text"
    />
    <input
      placeholder="Enter your annual salary"
      type="text"
    />
    <div>
      <button
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
